framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,706 @@
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method parameters' types.
+     */
+    public static void updateInferredMethodParametersTypes(
+            MethodInvocationNode methodInvNode, Tree receiverTree,
+            ExecutableElement methodElt, AnnotatedTypeFactory atf) {
+        if (receiverTree == null) {
+            // Method called from static context.
+            // I struggled to obtain the ClassTree of a method called from a
+            // static context and currently I'm ignoring it.
+            return;
+        }
+        ClassSymbol classSymbol = getEnclosingClassSymbol(receiverTree);
+        if (classSymbol == null) {
+            // Also struggled to obtain the ClassTree from an anonymous class.
+            // Ignoring it for now.
+            return;
+        }
+        String className = classSymbol.flatname.toString();
+
+        String jaifPath = getJaifPath(className);
+        AClass clazz = getAClass(className, jaifPath);
+
+        String methodName = JVMNames.getJVMMethodName(methodElt);
+        AMethod method = clazz.methods.vivify(methodName);
+
+        for (int i = 0; i < methodInvNode.getArguments().size(); i++) {
+            VariableElement ve = methodElt.getParameters().get(i);
+            AnnotatedTypeMirror paramATM = atf.getAnnotatedType(ve);
+
+            Node arg = methodInvNode.getArgument(i);
+            Tree treeNode = arg.getTree();
+            if (treeNode == null) continue;
+            AnnotatedTypeMirror argATM = atf.getAnnotatedType(treeNode);
+            AField param = method.parameters.vivify(i);
+            updateAnnotationSetInScene(
+                    param.type, atf, jaifPath, argATM, paramATM,
+                    DefaultLocation.PARAMETERS);
+        }
+    }
+
+    /**
+     * Updates the type of the field lhs in the Scene of the class with
+     * symbol classSymbol.
+     * <p>
+     * If the Scene contains no entry for the field lhs,
+     * the entry will be created and its type will be the type of rhs. If the
+     * Scene previously contained an entry/type for lhs, its new type will be
+     * the LUB between the previous type and the type of rhs.
+     * <p>
+     * @param lhs the field whose type will be refined.
+     * @param rhs the expression being assigned to the field.
+     * @param classTree the ClassTree where the assignment is invoked 
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the field's type.
+     */
+    public static void updateInferredFieldType(
+            Node lhs, Node rhs, ClassTree classTree, AnnotatedTypeFactory atf) {
+        FieldAccessNode lhsFieldNode = (FieldAccessNode) lhs;
+        ClassSymbol classSymbol = getEnclosingClassSymbol(classTree, lhsFieldNode);
+        if (classSymbol == null) return; // Anonymous class => Ignore, for now.
```

Comment:
Should this be marked with a TODO?
(This comment answers one of my previous questions and could be copied elsewhere.)